### PR TITLE
[3670] API Generation

### PR DESCRIPTION
**Pull request by wwelling**
**[VIVO GitHub issue](https://github.com/vivo-project/VIVO/issues)**: 
https://github.com/vivo-project/VIVO/issues/3670
https://github.com/vivo-project/VIVO/issues/3677

# What does this pull request do?
Create endpoints to generate OpenAPI specifications for an entire REST version, specific REST resource, all custom actions, or specific custom action as either application/yaml or application/json.

All REST yaml
https://editor.swagger.io/?url=https://gist.githubusercontent.com/wwelling/98775eeb6cac7c98fa90b24316d30711/raw/8a46f2e11d0f246957212f3b2e9cf079340ee1bb/vivo-generated-rest-api.yaml

All RPC yaml
https://editor.swagger.io/?url=https://gist.githubusercontent.com/wwelling/4109bfcc8d9cc6f10fcdcd983cc055d4/raw/509ea688227787e4513d36defc1b1e0db3a77ad9/vivo-generated-rpc-api.yaml

Single REST resource json
https://editor.swagger.io/?url=https://gist.githubusercontent.com/wwelling/76b5b2b7abb9d53137afbf684bcb803b/raw/997a393374f143f5711bdb62c131ce947d5cc830/test_person_resource.json

Single RPC action json
https://editor.swagger.io/?url=https://gist.githubusercontent.com/wwelling/b762b8ba097eb5fe96034c302a4e0c56/raw/39aad70e3353944e2e5d39e6f69c574dfd86d8f2/test_action.json

# What it is not doing

- Does not define the server object. https://swagger.io/specification/#server-object
- Does not include validations for parameters.
- Does not define error responses completely.
- Does not define any headers.
- Some missing descriptions as not available in RDF representation.
- Does not provide examples.

# Interested parties
Tag (@ mention) interested parties or, if unsure, @VIVO-project/vivo-committers

### CODE REVIEW

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java [hunk 1]
```diff
@@ -1,5 +1,770 @@
...
+        pathItem.setPost(actionPostOperation(action, tag));
+
+        return pathItem;
+    }
+
+    private Operation collectionGetOperation(Action action, Tag tag) {
+        Operation operation = new Operation();
+        operation.addTagsItem(tag.getName());
+
+        // No description available per resource API rpc
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:402]: Maybe it doesn't not exist in examples, but I believe it is allowed to be added in accordance with the ontology: https://github.com/vivo-project/Vitro/blob/sprint-dynapi-2022-feb-staging/home/src/main/resources/rdf/tbox/filegraph/vitro-dynamic-api.owl#L441

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java [hunk 2]
```diff
@@ -1,5 +1,770 @@
...
+
+        return operation;
+    }
+
+    private Operation collectionPostOperation(Action action, Tag tag) {
+        Operation operation = new Operation();
+        operation.addTagsItem(tag.getName());
+
+        // No description available per resource API rpc
+        // operation.setDescription("");
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:422]: See my previous comment

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java [hunk 3]
```diff
@@ -1,5 +1,770 @@
...
+
+        operation.setResponses(apiResponses);
+
+        return operation;
+    }
+
+    private PathParameter individualPathParameter() {
+        PathParameter pathParameter = new PathParameter();
+
+        pathParameter.setName("resourceId");
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:606]: maybe to define a constant, or to use this one https://github.com/vivo-project/Vitro/blob/sprint-dynapi-2022-feb-staging/api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/OperationData.java#L18 (you can change access to public and value to "resourceId").

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/request/ApiRequestPath.java [hunk 4]
```diff
@@ -7,7 +7,7 @@
 
 import javax.servlet.http.HttpServletRequest;
 
-public class RequestPath {
+public class ApiRequestPath {
 
     public static final String RESOURCE_ID_PARAM = "RESOURCE_ID";
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/request/ApiRequestPath.java:12]: Again RESOURCE_ID, this time capitalized, can we align all those constants?

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/request/DocsRequestPath.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.request;
+
+import static org.apache.commons.lang3.StringUtils.EMPTY;
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
+
+import javax.servlet.http.HttpServletRequest;
+
+import edu.cornell.mannlib.vitro.webapp.dynapi.request.ApiRequestPath.RequestType;
+
+public class DocsRequestPath {
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/request/DocsRequestPath.java:10]: Consider introducing an interface or abstract root for DocsRequestPath and ApiRequestPath?

**Review by chenejac — APPROVED** [review]
@wwelling  and @kaladay well done. I have added a couple of comments, please read it and check whether some of them are useful for some minor improvements before we merge this contribution.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java (hunk 1 above)
Comment (wwelling) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:402]: https://github.com/vivo-project/Vitro/blob/sprint-dynapi-2022-feb-staging/api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ResourceAPI.java

No description referenced in Java class.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java [hunk 6]
```diff
@@ -1,5 +1,770 @@
...
+        OpenAPI openApi = new OpenAPI();
+
+        APIInformation apiInformation = lookupApiInformation(requestPath);
+
+        if (apiInformation == null) {
+            log.error(format("Unable to find API information for version %s", requestPath.getApiVersion()));
+            return openApi;
+        }
+
+        System.out.print("Matched api (" + apiInformation.getVersion() + ") is " + apiInformation.getTitle() + ".");
```
[the remark this answers is not in the document]
Comment (wwelling) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:100]: Logs dont show up in tests. :(

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/request/DocsRequestPath.java (hunk 5 above)
Comment (wwelling) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/request/DocsRequestPath.java:10]: Deferring for future work.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java (hunk 1 above)
Comment (chenejac) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:403]: Actually, it should be added to https://github.com/vivo-project/Vitro/blob/sprint-dynapi-2022-feb-staging/api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/RPC.java, right? We need separated description for PUT, POST, GET and DELETE methods? Anyway, it is not there, and I will add it in the separated PR.

**Review by chenejac — APPROVED** [review]

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java [hunk 7]
```diff
@@ -1,5 +1,771 @@
...
+
+    private ConfigurationBeanLoader loader;
+    private ContextModelAccess modelAccess;
+    private OntModel dynamicAPIModel;
+
+    public static DynamicAPIDocumentation getInstance() {
+        return INSTANCE;
+    }
+
+    public void init(ServletContext ctx) {
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:73]: I believe this should be initialized in the [Initializer](https://github.com/vivo-project/Vitro/blob/sprint-dynapi-2022-feb-staging/api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/Initializer.java) similarly as Pools (Action and ResourceAPI).

File: api/pom.xml [hunk 8]
```diff
@@ -70,7 +70,13 @@
             <artifactId>fluent-hc</artifactId>
             <version>4.5.6</version>
         </dependency>
-        
+
+        <dependency>
```
Comment (chenejac) [inline api/pom.xml:74]: There is a conflict with jackson core 2.8.6 included as a dependency with orcid-api-client

**Review by chenejac — CHANGES_REQUESTED** [review]

**Review by chenejac — APPROVED** [review]

**Review by litvinovg — APPROVED** [review]
Looks and works good.

**Pull request merged**